cpp/src/arrow/util/heap.h

```diff
@@ -0,0 +1,76 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#pragma once
+
+#include <algorithm>
+#include <functional>
+#include <vector>
+#include "arrow/util/macros.h"
+
+namespace arrow {
+namespace internal {
+
+// A Heap class, is a simple wrapper to make heap operation simpler.
+// This class is immutable by design
+template <typename T, typename Compare = std::less<T>>
+class ARROW_EXPORT Heap {
+ public:
+  explicit Heap() : values_(), comp_() {}
```

Comment:
note there are some lints for this file: https://github.com/apache/arrow/pull/11019/checks?check_run_id=3468575773

```
 /arrow/cpp/src/arrow/util/heap.h:33:  Zero-parameter constructors should not be marked explicit.  [runtime/explicit] [5]
/arrow/cpp/src/arrow/util/heap.h:76:  Could not find a newline character at the end of the file.  [whitespace/ending_newline] [5]
```

Reply:
You can use docker to run linting locally: https://arrow.apache.org/docs/developers/cpp/development.html#code-style-linting-and-ci